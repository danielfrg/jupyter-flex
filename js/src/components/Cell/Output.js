import React from "react";

class CellOutput extends React.Component {
    state = { displayData: "" };

    componentDidMount() {
        // Display Priority

        let type = "";
        if ("image/svg+xml" in this.props.data) {
            type = "image/svg+xml";
        } else if ("image/png" in this.props.data) {
            type = "image/png";
        } else if ("text/html" in this.props.data) {
            type = "text/html";
        } else if ("text/markdown" in this.props.data) {
            type = "text/markdown";
        } else if ("image/jpeg" in this.props.data) {
            type = "image/jpeg";
        } else if ("text/plain" in this.props.data) {
            type = "text/plain";
        } else if ("text/latex" in this.props.data) {
            type = "text/latex";
        } else if ("application/javascript" in this.props.data) {
            type = "application/javascript";
        } else if (
            "application/vnd.jupyter.widget-state+json" in this.props.data
        ) {
            type = "application/vnd.jupyter.widget-state+json";
        } else if (
            "application/vnd.jupyter.widget-view+json" in this.props.data
        ) {
            type = "application/vnd.jupyter.widget-view+json";
        }

        this.setState({ displayData: type });
    }

    createMarkup(html) {
        return { __html: html };
    }

    runScript(reportScript) {
        // This runs the contents in script tag on a window/global scope
        let scriptToRun = reportScript;
        if (scriptToRun !== undefined) {
            //remove <script> and </script> tags since eval expects only code without html tags
            let scriptLines = scriptToRun.split("\n");
            scriptLines.pop();
            scriptLines.shift();
            let cleanScript = scriptLines.join("\n");
            // console.log("running script ", cleanScript);
            window.eval(cleanScript);
        }
    }

    displayHTML(data) {
        // What can go wrong
        // https://stackoverflow.com/questions/35614809/react-script-tag-not-working-when-inserted-using-dangerouslysetinnerhtml

        const content = data["text/html"];
        let extractedScript = /<script[\s\S]*<\/script>/g.exec(content);
        if (extractedScript) {
            this.runScript(extractedScript[0]);
        }
        return (
            <div
                className="output_html rendered_html output_subarea"
                dangerouslySetInnerHTML={this.createMarkup(data["text/html"])}
            ></div>
        );
    }

    displayTextPlain(data) {
        return <pre>{data["text/plain"]}</pre>;
    }

    displaySVG(data) {
        return (
            <div
                className="jp-RenderedSVG jp-OutputArea-output"
                data-mime-type="image/svg+xml"
                dangerouslySetInnerHTML={this.createMarkup(
                    data["image/svg+xml"]
                )}
            ></div>
        );
    }

    getMetadata(key, mimetype = "") {
        if (mimetype && mimetype in this.props.metadata) {
            return this.props.metadata[mimetype][key];
        }
        return this.props.metadata[key];
    }

    displayPNG(data) {
        const { metadata } = this.props;

        let params;
        const width = this.getMetadata("width", "image/png");
        const height = this.getMetadata("height", "image/png");
        const class_ = this.getMetadata("unconfined", "image/png");
        if (width) {
            params["width"] = width;
        }
        if (height) {
            params["height"] = height;
        }
        if (class_) {
            params["className"] = class_;
        }

        const usesFilesnames = "filenames" in metadata ? true : false;

        let components;
        if (usesFilesnames) {
            components = (
                <img src={metadata["filenames"]["image/png"]} {...params}></img>
            );
        } else {
            components = (
                <img
                    src={`data:image/png;base64,${data["image/png"]}`}
                    {...params}
                ></img>
            );
        }

        return (
            <div className="jp-RenderedImage jp-OutputArea-output">
                {components}
            </div>
        );
    }

    displayJPEG(data) {
        const { metadata } = this.props;

        let params;
        const width = this.getMetadata("width", "image/jpeg");
        const height = this.getMetadata("height", "image/jpeg");
        const class_ = this.getMetadata("unconfined", "image/jpeg");
        if (width) {
            params["width"] = width;
        }
        if (height) {
            params["height"] = height;
        }
        if (class_) {
            params["className"] = class_;
        }

        const usesFilesnames = "filenames" in metadata ? true : false;

        let components;
        if (usesFilesnames) {
            components = (
                <img
                    src={metadata["filenames"]["image/jpeg"]}
                    {...params}
                ></img>
            );
        } else {
            components = (
                <img
                    src={`data:image/jpeg;base64,${data["image/jpeg"]}`}
                    {...params}
                ></img>
            );
        }

        return (
            <div className="jp-RenderedImage jp-OutputArea-output">
                {components}
            </div>
        );
    }

    render() {
        const { data, metadata, output_type } = this.props;
        let el = "";

        switch (this.state.displayData) {
            case "image/svg+xml":
                el = this.displaySVG(data);
                break;
            case "image/png":
                el = this.displayPNG(data);
                break;
            case "text/html":
                el = this.displayHTML(data);
                break;
            case "text/markdown":
                break;
            case "image/jpeg":
                el = this.displayJPEG(data);
                break;
            case "text/plain":
                el = this.displayTextPlain(data);
                break;
            case "text/latex":
                break;
            case "application/javascript":
                break;
            case "application/vnd.jupyter.widget-state+json":
                break;
            case "application/vnd.jupyter.widget-view+json":
                break;
        }

        return el;
    }
}

export default CellOutput;
