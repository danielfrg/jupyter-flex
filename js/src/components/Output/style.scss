.card {
    .output {
        pre {
            margin: 8px 0;
        }

        table.dataframe {
            border: none;
            border-collapse: collapse;
            border-spacing: 0;
            color: #000;
            font-size: 12px;
            table-layout: fixed;
            width: 100%;

            thead {
                border-bottom: 1px solid #000;
                vertical-align: bottom;
            }
            tr,
            th,
            td {
                text-align: right;
                vertical-align: middle;
                padding: 0.5em 0.5em;
                line-height: normal;
                white-space: normal;
                max-width: none;
                border: none;
            }
            th {
                font-weight: bold;
            }
            tbody tr:nth-child(odd) {
                background: #f5f5f5;
            }
            tbody tr:hover {
                background: rgba(66, 165, 245, 0.2);
            }
            * + table {
                margin-top: 1em;
            }
        }
    }

    .card-footer {
        p {
            font-size: 12px;
        }
        p:last-child {
            margin-bottom: 0;
        }
    }
}
