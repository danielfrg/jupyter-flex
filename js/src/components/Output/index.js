import React from "react";
import Convert from "ansi-to-html";

import { createMarkup, uuidv4, onNextFrame } from "../utils";

import "./style.scss";

export function runScript(script) {
    // This runs the contents in script tag on a window/global scope

    // We do that weird closing script tag replace so it works in dev mode...
    let reportScript = script.trim();
    if (
        reportScript.startsWith("<script") &&
        reportScript.endsWith("< /script>".replace(" ", ""))
    ) {
        // remove opening and closing script tags
        // windows.eval expects code without html tags
        let scriptLines = reportScript.split("\n");
        scriptLines.pop();
        scriptLines.shift();
        reportScript = scriptLines.join("\n");
    }
    if (reportScript) {
        // Ok this part really is the worst lol
        // This is a hack to make Altair code not look for previousElementSibling
        // because of how we execute the script `document.currentScript` is always null
        // The script from altair has a fallback to look for the right ID
        // So we are always triggering that fallback
        // I should make a PR to vega or something to make this more secure
        reportScript = reportScript.replace(
            "let outputDiv = document.currentScript.previousElementSibling;",
            "let outputDiv = {id: 0};"
        );
        // console.log("running script:");
        // console.log(reportScript);
        window.eval(reportScript);
    }
}

class Output extends React.Component {
    ansiConverter;
    state = { displayData: "" };

    constructor(props) {
        super(props);
        this.ansiConverter = new Convert();

        let type = "";
        // if ("application/vnd.bokehjs_exec.v0+json" in this.props.metadata) {
        //     type = null;
        // } else
        if (this.props.data) {
            if ("image/svg+xml" in this.props.data) {
                type = "image/svg+xml";
            } else if ("image/png" in this.props.data) {
                type = "image/png";
            } else if ("text/html" in this.props.data) {
                type = "text/html";
            } else if ("text/markdown" in this.props.data) {
                type = "text/markdown";
            } else if ("image/jpeg" in this.props.data) {
                type = "image/jpeg";
            } else if ("text/latex" in this.props.data) {
                type = "text/latex";
            } else if ("application/javascript" in this.props.data) {
                type = "application/javascript";
            } else if (
                "application/vnd.jupyter.widget-state+json" in this.props.data
            ) {
                type = "application/vnd.jupyter.widget-state+json";
            } else if (
                "application/vnd.jupyter.widget-view+json" in this.props.data
            ) {
                type = "application/vnd.jupyter.widget-view+json";
            } else if ("text/plain" in this.props.data) {
                type = "text/plain";
            }
        }

        this.state = { displayData: type };
    }

    componentDidMount() {
        if (this.state.displayData == "text/html") {
            // If there are any script tags in the HTML code then we need to execute those
            // We need to do it after the components have been writen to the DOM
            // Also we need to use the runScript() util to run the code directly into the DOM
            // Since react doesn't execute this tags.
            // https://stackoverflow.com/questions/35614809/react-script-tag-not-working-when-inserted-using-dangerouslysetinnerhtml

            const content = this.props.data["text/html"];
            let extractedScript = /<script[\s\S]>*<\/script>/g.exec(content);
            if (extractedScript) {
                onNextFrame(() => {
                    runScript(extractedScript[0]);
                });
            }
        } else if (this.state.displayData == "application/javascript") {
            runScript(this.props.data["application/javascript"]);
        }
    }

    render() {
        const { data, output_type } = this.props;

        if (output_type == "error") {
            return this.displayError(this.props.traceback);
        }
        if (output_type == "stream") {
            return this.displayStream(this.props.text);
        }

        let el = "";
        switch (this.state.displayData) {
            case "image/svg+xml":
                el = this.displaySVG(data);
                break;
            case "image/png":
                el = this.displayPNG(data);
                break;
            case "text/html":
                el = this.displayHTML(data);
                break;
            case "text/markdown":
                break;
            case "image/jpeg":
                el = this.displayJPEG(data);
                break;
            case "text/plain":
                el = this.displayTextPlain(data);
                break;
            case "text/latex":
                break;
            case "application/javascript":
                el = this.displayJavascript(data);
                break;
            case "application/vnd.jupyter.widget-state+json":
                el = this.displayWidgetState(data);
                break;
            case "application/vnd.jupyter.widget-view+json":
                el = this.displayWidgetView(data);
                break;
        }

        return el;
    }

    getMetadata(key, mimetype = "") {
        if (mimetype && mimetype in this.props.metadata) {
            return this.props.metadata[mimetype][key];
        }
        return this.props.metadata[key];
    }

    displayTextPlain(data) {
        return <pre>{data["text/plain"]}</pre>;
    }

    displayHTML(data) {
        return (
            <div
                className="output_html rendered_html output_subarea"
                dangerouslySetInnerHTML={createMarkup(data["text/html"])}
            ></div>
        );
    }

    displayJavascript(data) {
        return <script>{data["application/javascript"]}</script>;
    }

    displaySVG(data) {
        return (
            <div
                className="jp-RenderedSVG jp-OutputArea-output"
                data-mime-type="image/svg+xml"
                dangerouslySetInnerHTML={createMarkup(data["image/svg+xml"])}
            ></div>
        );
    }

    displayPNG(data) {
        const { metadata } = this.props;

        let params;
        const width = this.getMetadata("width", "image/png");
        const height = this.getMetadata("height", "image/png");
        const class_ = this.getMetadata("unconfined", "image/png");
        if (width) {
            params["width"] = width;
        }
        if (height) {
            params["height"] = height;
        }
        if (class_) {
            params["className"] = class_;
        }

        const usesFilesnames = "filenames" in metadata ? true : false;

        let components;
        if (usesFilesnames) {
            components = (
                <img
                    src={metadata["filenames"]["image/png"]}
                    alt="Output"
                    {...params}
                ></img>
            );
        } else {
            components = (
                <img
                    src={`data:image/png;base64,${data["image/png"]}`}
                    alt="Output"
                    {...params}
                ></img>
            );
        }

        return (
            <div className="jp-RenderedImage jp-OutputArea-output">
                {components}
            </div>
        );
    }

    displayJPEG(data) {
        const { metadata } = this.props;

        let params;
        const width = this.getMetadata("width", "image/jpeg");
        const height = this.getMetadata("height", "image/jpeg");
        const class_ = this.getMetadata("unconfined", "image/jpeg");
        if (width) {
            params["width"] = width;
        }
        if (height) {
            params["height"] = height;
        }
        if (class_) {
            params["className"] = class_;
        }

        const usesFilesnames = "filenames" in metadata ? true : false;

        let components;
        if (usesFilesnames) {
            components = (
                <img
                    src={metadata["filenames"]["image/jpeg"]}
                    alt="Output"
                    {...params}
                ></img>
            );
        } else {
            components = (
                <img
                    src={`data:image/jpeg;base64,${data["image/jpeg"]}`}
                    alt="Output"
                    {...params}
                ></img>
            );
        }

        return (
            <div className="jp-RenderedImage jp-OutputArea-output">
                {components}
            </div>
        );
    }

    displayWidgetState(data) {
        const uuid = uuidv4();
        return (
            <div id={uuid} className="output_subarea output_widget_state">
                <script type="text/javascript">
                    {`var element = document.getElementById("${uuid}");`}
                </script>
                <script type="application/vnd.jupyter.widget-state+json">
                    {JSON.stringify(
                        data["application/vnd.jupyter.widget-state+json"]
                    )}
                </script>
            </div>
        );
    }

    displayWidgetView(data) {
        const uuid = uuidv4();
        return (
            <div id={uuid} className="output_subarea output_widget_state">
                <script type="text/javascript">
                    {`var element = document.getElementById("${uuid}");`}
                </script>
                <script type="application/vnd.jupyter.widget-view+json">
                    {JSON.stringify(
                        data["application/vnd.jupyter.widget-view+json"]
                    )}
                </script>
            </div>
        );
    }

    displayError(text) {
        return (
            <div
                className="output_html rendered_html output_subarea"
                dangerouslySetInnerHTML={createMarkup(
                    this.ansiConverter.toHtml(text)
                )}
            ></div>
        );
    }

    displayStream(text) {
        return (
            <div className="output_subarea output_stream output_stdout output_text">
                <pre>{this.ansiConverter.toHtml(text)}</pre>
            </div>
        );
    }
}

export default Output;
