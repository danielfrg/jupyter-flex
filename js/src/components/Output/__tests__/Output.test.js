import React from "react";
import { create } from "react-test-renderer";
import { test, expect } from "jest";

import Output from "../../Output";

test("empty", () => {
    const c = create(<Output />);
    expect(c.toJSON()).toMatchSnapshot();
});

// test("png", () => {
//     const c = create(<Output data={} />);
//     expect(c.toJSON()).toMatchSnapshot();
// });
