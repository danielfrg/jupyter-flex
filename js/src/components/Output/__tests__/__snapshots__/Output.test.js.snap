// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty 1`] = `""`;
